Ensure backbeat retrieve the account quotas

williamlardier opened the pull request:
Fix for Quota during expiration

Issue: CLDSRV-590

Seragonia committed:
Ensure backbeat retrieve the account quotas

The route backbeat also calls the doAuth function, but we never
map the value to accountQuotas

Issue: CLDSRV-590

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-590 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.39`


Please check the `Fix Version/s` of CLDSRV-590, or the target
branch of this pull request.

Seragonia committed:
Release 8.8.39

Issue: CLDSRV-590

ghost commented:
/approve

Seragonia committed:
Ensure we pass the right action name for delete object APIs

From backbeat routes, all APIs will use a name we only need for
monitoring, but the action name is actually used during the
quota evalutation (and possibly bucket policy as well).

Issue: CLDSRV-590

francoisferrand commented on lib/api/objectDelete.js:
```diff
@@ -51,6 +51,13 @@ function objectDeleteInternal(authInfo, request, log, isExpiration, cb) {
+    // This is also needed for quota evaluation, where "routeBackbeat" cannot
+    // be used for different operations, such as putObject
+    // eslint-disable-next-line no-param-reassign
+    request.apiMethod = reqVersionId ? 'objectDeleteVersion' : 'deleteObject';
```
would it not be better or safer to keep the existing `apiMethod`, and just pass a different `requestType` in case of `isExpiration` ?
→ this ensures only `valParams` is changed, i.e. only the behavior of `standardMetadataValidateBucketAndObj()` is affected
→ it explicitely tweaks the behavior for the backbeat route only, ensuring the "user" api/path is not affected

francoisferrand replied:
instead of the "generic" objectDelete, we do this only in `deleteObjectFromExpiration` :

```suggestion
function deleteObjectFromExpiration(request, response, userInfo, log, callback) {
    // Reassign a specific apiMethod, as it is needed for quota evaluation (at least), where
    // "routeBackbeat" cannot be used as it is used for all backbeat API operations...
    // eslint-disable-next-line no-param-reassign
    request.apiMethod = 'objectDelete';
```

francoisferrand commented on lib/api/objectDelete.js:
```diff
@@ -51,6 +51,13 @@ function objectDeleteInternal(authInfo, request, log, isExpiration, cb) {
     const reqVersionId = decodedVidResult;
     const hasGovernanceBypass = hasGovernanceBypassHeader(request.headers);
 
+    // Reassign a specific apiMethod, as monitoring can rely on the
```
as far as I can tell (looking at the code), `apiMethod` is not used for monitoring: but mostly to
* override the default `requestType` passed to `standardMetadataValidateBucketAndObj()` in every api....
* in prepareRequestContexts().... but not for backbeat routes, which do the evaluation by assuming an api method of 'objectReplicate' 😱
* in the "standard" routes, to handle special handling of specific APIs (which may be needed for backbeat routes as welll...fortunately we don't seem to use these affected API -mostly `GetObject`/`PutObject`- through backbeat routes)

francoisferrand commented on lib/routes/routeBackbeat.js:
```diff
@@ -1388,7 +1388,7 @@ function routeBackbeat(clientIP, request, response, log) {
         const { host, port } = config.backbeat;
         const target = `http://${host}:${port}${path}`;
-        return auth.server.doAuth(request, log, (err, userInfo) => {
+        return auth.server.doAuth(request, log, (err, userInfo, streamingV4Params, infos) => {
```
not the right one to change: this is used when forwarding to backbeat api pod...

the auth call for "cloudserver" backbeat routes is below... (line ~1465)

francoisferrand committed:
Track inflights for backbeat's batchDelete route

This is needed to track deletions in case of transition or restored
object expiration.

Issue: CLDSRV-590

williamlardier commented on lib/routes/routeBackbeat.js:
```diff
@@ -1176,7 +1178,24 @@ function batchDelete(request, response, userInfo, log, callback) {
-            return _respond(response, null, log, callback);
+
+            // Update inflight metrics for the data which has just been freed
+            const bucket = parsedPayload.Bucket;
```
this part LGTM

Seragonia committed:
Bump Arsenal 8.1.143

Issue: CLDSRV-590

Seragonia committed:
Ensure we follow the right return variables from doAuth

BB routes are not authorizing internal API calls. To
be adressed in CLDSRV-591

Issue: CLDSRV-590

francoisferrand committed:
Fix quota inflights for cold objects

Deleting cold objects should not increment inflight, unless they are
restored.

Issue: CLDSRV-590

francoisferrand committed:
Fix inflight handling for multiObjectDelete

Issue: CLDSRV-590

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.8`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`



Please check the status of the associated issue CLDSRV-590.

Goodbye williamlardier.




*The following options are set:* **approve**
